btf/traversal.go

```diff
@@ -0,0 +1,52 @@
+package btf
+
+// walkType calls fn on each child of typ.
+//
+// It's faster than directly invoking typ.children for some common types.
+func walkType(typ Type, fn func(*Type)) {
```

Comment:
Since this now contains a type switch, wouldn't this technically allow us to remove `children()` from interface `Type` and `walk(nil)` in the default case?

That would also allow us to tailor method names to the receiver, e.g. `Pointer.children()` could become `Pointer.target()` and `FuncProto.params()`, etc. Just a naive thought from looking at it from a distance.

From a domain definition perspective, a `children()` method doesn't make much sense as the method of traversal is so highly specific to the type and is hard to define in a generic way.

Reply:
Yes, that would be possible, that's kind of what this function started out as. The downside is that it's quite easy to forget to extend walkType when adding a new implementation of Type, since there is no compiler feedback. It just blows up at some point.

Maybe not such a big deal, since we rarely add new types and also never want to allow users to implement Type. Should I change it?

Reply:
> It just blows up at some point.

Well, my guess would be the opposite. If `walkType` encounters an unknown type, the default clause will cause traversal to silently stop, which is always more difficult to debug than a hard error.

An interface method reminds us to define traversal behaviour when adding a new Type, though in the current iteration with the default clause, it's still possible to implement a Type and forget to add it to the type switch, which will incur a heap alloc when traversing that particular Type.

Reply:
I've looked at linters for this, nothing existing fits the bill. Some combination of https://github.com/nishanths/exhaustive/blob/master/enum.go and https://github.com/dominikh/implements/blob/master/main.go could work.

Some redeeming factors:
- If a new type shows up in BTF in the wild, I assume things will fail early on in the pipeline? (e.g. during inflation/parsing?)
- Reminder to extend `walkType` should be documented on the `children()` interface method.

Then I think we should be good.

Reply:
I pushed a change. `children()` is now gone since there are no other callers left, and it's just duplicate code. Please take a look!